### ISSUE TITLE
gradlew and `--no-daemon` result in unexpected JVM fork

### ISSUE THREAD

**Issue by rocketraman** [gradle/gradle#16110]
### Expected Behavior
When using the `--no-daemon` flag, my expectation is that if no JVM settings are defined e.g. (`org.gradle.jvmargs`), no JVM will be forked i.e. `GradleMain` will start rather than `GradleDaemon`.

### Current Behavior
The `gradlew` script contains `DEFAULT_JVM_OPTS='"-Xmx64m" "-Xms64m"'`. The existence of this causes Gradle to fork a JVM, unless the script is edited to remove those default settings.

### Context
I had a problem with a Kotlin native build. It consistently causes the exception:

```
org.gradle.launcher.daemon.client.DaemonDisappearedException: Gradle build daemon disappeared unexpectedly (it may have been killed or may have crashed)
```
and the daemon log contains no useful information. Only when running the build non-forked can I see the true error output (the fact that Gradle does not show this output when forked is probably a different bug):

```
error: Linking globals named 'cocoapods_FirebaseAuth_FIRAuthErrorDomain_getter_wrapper0': symbol multiply defined!
```

I was unable to see this error until I finally finagled Gradle into not forking.

### Steps to Reproduce 
Simply run `./gradlew --no-daemon build`.

### Your Environment
```
------------------------------------------------------------
Gradle 6.8.2
------------------------------------------------------------

Build time:   2021-02-05 12:53:00 UTC
Revision:     b9bd4a5c6026ac52f690eaf2829ee26563cad426

Kotlin:       1.4.20
Groovy:       2.5.12
Ant:          Apache Ant(TM) version 1.10.9 compiled on September 27 2020
JVM:          11.0.9.1 (AdoptOpenJDK 11.0.9.1+1)
OS:           Mac OS X 10.16 x86_64
```

**Comment by asarkar** [comment]
> Only when running the build non-forked

How do you do that, by editing `gradlew` script and removing `DEFAULT_JVM_OPTS`?

**Comment by rocketraman** [comment]
> How do you do that, by editing `gradlew` script and removing `DEFAULT_JVM_OPTS`?

Yes, set it to an empty value.

**Comment by rocketraman** [comment]
Not stale

**Comment by gerhard6** [comment]
quite inconvenient, that we need to modify gradlew to achieve the intended effect - we have some startup issues with the daemon in CI, and preventing the daemon could perhaps help us.
Same problem on gradlew.bat, btw.

**Comment by gerhard6** [comment]
> quite inconvenient, that we need to modify gradlew to achieve the intended effect - we have some startup issues with the daemon in CI, and preventing the daemon could perhaps help us. Same problem on gradlew.bat, btw.

seems that modifying gradlew is not necessary to prevent the daemon from starting: just using  eg "JAVA_OPTS=-Xmx512m" works. This overrides the DEFAULT_JVM_OPTS

**Comment by rocketraman** [comment]
Actually, when explicitly setting `JAVA_OPTS` if there is no `org.gradle.jvmargs` does appear to work around the issue somehow, BUT that appears to break the opposite case -- if `org.gradle.jvmargs` *is set*, then setting any value (or no value) for `JAVA_OPTS` forks again, even if `JAVA_OPTS` is set to the *same* value as `org.gradle.jvmargs`:

```
$ grep org.gradle.jvmargs ~/.gradle/gradle.properties
org.gradle.jvmargs=-Xms2048M -Xmx4096M -XX:MaxMetaspaceSize=1g


```
./gradlew --no-daemon compileKotlin
To honour the JVM settings for this build a single-use Daemon process will be forked. See https://docs.gradle.org/7.4/userguide/gradle_daemon.html#sec:disabling_the_daemon.
[...snip...]

JAVA_OPTS="-Xms2048M -Xmx4096M -XX:MaxMetaspaceSize=1g" ./gradlew --no-daemon compileKotlin
To honour the JVM settings for this build a single-use Daemon process will be forked. See https://docs.gradle.org/7.4/userguide/gradle_daemon.html#sec:disabling_the_daemon.
[...snip...]
```

Basically this is very messy. The idea should be if I specify `--no-daemon` I don't want a daemon! Gradle should respect that and warn me if the JVM settings are not *respected* i.e.


```
The user requested no daemon. The JVM settings `-X...` cannot be honored for this build as the current JVM settings are `-X...`. See https://docs.gradle.org/7.4/userguide/gradle_daemon.html#sec:disabling_the_daemon.
```

**Comment by ljohnston** [comment]
This needs to be addressed. It shouldn't be this hard to turn off the daemon.

**Comment by robert-dodier** [comment]
For the record, I've bumped into this issue and it would be great to address it somehow. I mention it just in case it makes a difference how many people are seeing this problem.

**Comment by ersen-lw** [comment]
I'm having the same issue. I've attempted to make GRADLE_OPTS and JAVA_OPTS match but still results in a Daemon being made.

https://discuss.gradle.org/t/need-help-trying-to-stop-daemon-and-re-use-launcher-jvm/47852

**Comment by Fleshgrinder** [comment]
The `xargs` based logic breaks if a system property is set that contains a line feed, like `line.separator`. This not only means that a new JVM is being forked, but that it's impossible to actually set all system properties to avoid it.

**Comment by byteflys** [comment]
my probelem is, i need to debug gradle script
org.gradle.debug=true
org.gradle.daemon=false
debug need close daemon, but daemon cannot be closed
offical answer is, jvm args should match JAVA_OPTS/GRADLE_OPTS
but how to match, is really a difficult problem
if this probem is not well resolved
means most developers cannot debug their script and plugin

**Comment by ljacomet** [comment]
Given #29307, providing an update here:
* The current message could be improved to help users figure out what is not aligned.
* However, we also have #24057 

Based on this, the most likely evolution, that we are discussing for Gradle 9, is to deprecate and remove `--no-daemon` in favour of a flag that will allow you to say the daemon must expire after serving one build.

**Comment by robert-dodier** [comment]
@ljacomet wrote: "Based on this, the most likely evolution, that we are discussing for Gradle 9, is to deprecate and remove --no-daemon in favour of a flag that will allow you to say the daemon must expire after serving one build." 

That development path doesn't fix any of the underlying issues that people are running into, as reported here. People are wanting to run without creating a daemon for various reasons (although mostly having to do with greater comprehensibility and predictability of results). The business about trouble with the --no-daemon option is just some incidental fallout: it won't help any to simply obliterate the --no-daemon option and make it impossible to run without creating a daemon. It doesn't help users any to make the daemon expire after serving one build; the trouble is in the one build that it is serving.

**Comment by mikaelstaldal** [comment]
If you run Gradle in a Docker container (which is common in CI environments), it is annoying and inefficient if it always starts an additional JVM process for no good reason.

**Comment by sid-hbm** [comment]
This deamon thing is horrible and very dangerous in my opinion. The use case I have is very simple: I want to run gradle only once and I want to do very fast and be done. Launching the daemon takes long time and that defeats the purpose or running only once and fast. The option --no-daemon is not helping because all it does is run the daemon once and remove it, but that defeats the purpose of not wanting the daemon to run at all. I even tried so many other options like AVA_OPTS and GRADLE_OPTS match org.gradle.jvmargs, adding org.gradle.daemon=false to gradle.properties, etc... No matter what I do, still the daemon will be launch and take forever along with that annoying message "To honour the JVM settings for this build a single-use Daemon process will be forked". This is very critical of an issue. Maven was starting to become outdated, then Gradle came with a bigger shining light that you would think maven will be killed and gone, and now maven is doing much better in terms features and less issues, while Gradle that supposed to be best of the best is having this major issue that nobody cares to fix.

**Comment by meavydev** [comment]
We hit this problem due to ESETs killing off the gradle daemon:
 https://forum.eset.com/topic/41990-eset-block-gradle-after-latest-update/?do=findComment&comment=189817

So single use gradle daemon will not work, we actually need no gradle daemon, or we can't build at all.

**Comment by rocketraman** [comment]
I [wrote](https://github.com/gradle/gradle/issues/16110#issuecomment-2140689842):

> However, in doing so, please collect and consider all use-cases for the --no-daemon option, and plan to address the relevant ones.

Another use case for `--no-daemon` is development tools that modify a processes syscalls in order to provide a development capability. An example of this is [mirrord](https://mirrord.dev/docs/overview/introduction/), which allows a local process to connect to and participate in a remote Kubernetes cluster. To do so, it intercepts network calls and redirects them to the cluster.

When the actual JVM process is executed in a forked daemon, this mechanism does not work.

The workaround is to run the process outside of Gradle, but this adds overhead to the DX.

**Comment by cuzfrog** [comment]
Tried above workaround, didn't work for me, version is 8.14.3, using `gradlew`. In order to easily debug the build process (plugin code), I need to run in the same jvm. Though this is done by `-Dorg.gradle.debug=true`, still, hope this can be addressed. It's confusing and annoying.

**Comment by numinit** [comment]
We have always just executed GradleMain in nixpkgs and are discussing whether to remove the JVM memory opts to reproduce the more sensible behavior of avoiding the daemon with --no-daemon: https://github.com/NixOS/nixpkgs/pull/449037

**Comment by ljacomet** [comment]
This is resolved in Gradle 9.5.0 through the following issue:
* #33212


With info level logging, you will have the necessary information to make sure no forking happens, while it remains supported, see:
* #2660
